Speeding up the first phase of Index Building

joka921 opened the pull request:
- Identified the writing of Triple elements to hash maps for deduplication as a bottleneck.
- Speed this up by writing to multiple hash maps at once and merging them afterwards
- Also concurrently pipelined all the other steps in the first index building phase.
- That way the TurtleParsing now becomes the bottleneck (Teaser: This can be sped up by 30% using
  compile time regexes, so this might become even faster in a later PR).

- This was done by Implementing an abstract, templated BatchedPipeline that abstracts
  over the creation and transformation of values in a pipeline that allows to control
  the degree of concurrency used on each level and between the different levels.

- This pipeline infrastructure was heavily unit-tested to ensure its correctness since it internally
  uses quite some template magic.

- This Commit also introduces the absl::flat_hash_map that is faster than the dense_hash_map used before.
  But in doubt I can also revert this change since in the meantime the parallelism is what helps us more
  than the faster hash maps. But I have thought, that we wanted to try out those anyway.

- This can already be reviewed, especially the BatchedPipeline.h file. Before merging I would suggest merging
  the Unicode PR, because there is some merging work to be done (but not too much) between those PRs.

hannahbast commented:
@joka921 That sounds amazing! I have a few questions:

1. Did you already try it? What is the speedup?

2. Can you describe in a little more detail what you mean by "writing to multiple hash maps at once". We are talking about hash maps from strings (names) to IDs, right? How many hash maps? And how do you decide which hash map to use for a particular element?

3. Can you describe in a littel more detail, which operations you pipeline? You mention "values" and "levels", what are those when you use your abstract class for building the index?

4. Does the absl::flat_hash_map have any disadvantages compared to Google's dense_hash_map?

joka921 commented:
Ok, so for each triple we have to do the following steps:
1. Parse it from Turtle
2. Convert each of subject, predicate, object to an internal representation (e.g. string normalization, special handling of numeric or date literals etc.)
3. Assign each of subject, predicate, object an ID (lookup in a hash map if this element already has an ID, then reuse it, otherwise assign the next available ID and store it in the hash map).
4. Append the Triple's Ids to a large external vector that stores all the triples.

In a pipeline this looks as follows:
1. Parse a batch of Triples and pass this batch to the next level. Immediately continue with next batch.
2. Take a batch of Triples from level 1 and convert each element to the representation, then pass this
    converted batch to level 3. Immediately continue with next batch
3. The first 25% of each batch of Triples get Ids from the first hash map, the second 25% from the second.... 
4. get all the Ids from the different hashMaps and store the Id Triples.

Trick a) all of those levels happen at the same time for different batches (pipeline principle)
Trick b) We make sure that the hashMaps hand out disjoint sets of Ids (this is easy because we know  the maximum number of triples we deal with at once in this whole process.

After we are finished, we have the problem, that some words may have multiple Ids (if they occured in different triples that were assigned to different hash maps). Unifying these to one Id and updating the triple after we are done
is again relatively simple and can be done concurrently for each partial vocabulary.

The speed up is s.t. the parser becomes the bottle neck. It is about 40 Million triples per Minute so
we should be able to deal with 1 Billion triples in about 4 -5 hours vs 10-12 before. My personal target is 3, but this requires more work on the parser.

The only thing the `abls::flat_hash_map` does not have is a bucket interface (access to linked lists for chained hashing). This gains speed but does not conform to the C++ standard for `unordered_map`. But since nobody uses this, especially not us this is not a disadvantage. Concerning memory consumption both are rather expensive (optimized for speed) but since we currently use only relatively small hash maps (in my case they are even temporary and only used in the IndexBuild) this is not an issue for us.

hannahbast commented:
@joka921 Thanks for the explanation!

You write that "We make sure that the hashMaps hand out disjoint sets of Ids (this is easy because we know the maximum number of triples we deal with at once in this whole process.)". Does this mean that with this PR there are gaps in the ID space? Or do these gaps vanish in the ID unification process.

I am slightly worried about code complexity and maintainability. Do I understand you correctly that all the complexity is hidden in the BatchedPipeline class? Is the code outside that class reasonably simple. For example, if someone else wants to extend the parser in some (not overly dramatic way), will they be able to do it without understanding the intricacies of your batched pipeline?

joka921 commented:
1. No, we will not have Gaps in the Id space. The procedure described above does merge the Ids in a gap free way. Additionally it only creates a partial vocabulary. Those are merged after the complete parsing of the whole knowledge base is done. Only there the Ids are really finalized in a gap free way.
Anything else would be very wrong since the Id of a string must be its index in the vocabulary vector when we are completely finished.

2. Using the batched pipeline is really really simple, and I have written quite some documentation.
    The Parser implementation is completely independent, as long as the parser can give us triples
    and signal when the file is finished and there are more triples.

3. The interface of the pipeline is basically `setupPipeline(one Lambda per Step)`
    You can look at it yourself, or we can meet some time soon and discuss on how to comment this
    best and make the actual usage readable in the best possible way.

joka921 commented:
```
   {
      auto p = ad_pipeline::setupParallelPipeline<1, NUM_PARALLEL_ITEM_MAPS>(
          _parserBatchSize,
          [parserPtr = &parser, i = 0ull, linesPerPartial,
           &parserExhausted]() mutable -> std::optional<Triple> {
            if (i >= linesPerPartial) {
              return std::nullopt;
            }
            Triple t;
            if (parserPtr->getLine(t)) {
              i++;
              return std::optional(std::move(t));
            } else {
              parserExhausted = true;
              return std::nullopt;
            }
          },
          [this](Triple&& t) {
            return tripleToInternalRepresentation(std::move(t));
          },
          std::move(itemMapLambdaTuple));

      while (auto opt = p.getNextValue()) {
        i++;
        for (const auto& innerOpt : opt.value()) {
          if (innerOpt) {
            actualCurrentPartialSize++;
            localWriter << innerOpt.value();
          }
        }
```

joka921 commented:
@niklas88 The Travis build says that it has passed when I click on the `Details` link, But
on this site the yellow light remains. Does the CI pipeline have a hickup again?

niklas88 commented:
> @niklas88 The Travis build says that it has passed when I click on the `Details` link, But
> on this site the yellow light remains. Does the CI pipeline have a hickup again?

I just reran the build and this time its propagated correctly. So yeaah looks like a Travis hickup. I'll look into this in more detail soon but it already sounds pretty amazing!

hannahbast commented on src/index/Index.cpp:
```diff
@@ -121,107 +124,123 @@ template void Index::createFromFile<TurtleMmapParser>(const string& filename);
+        if (i % 10000000 == 0) {
+          LOG(INFO) << "Lines (from KB-file) processed: " << i << '\n';
+        }
       }
```
@joka921 Could you add a few well-placed and concise comments in this code block so that the structure becomes clearer and what the individual lambdas do?

niklas88 replied:
I think you mentioned that you could also replace the lambdas with well named functions. I think this might indeed be helpful here

hannahbast commented:
I just tried to build a version of the current master (includes the unicode upgrade) with this PR merged locally and encountered this error while building the docker container:

```
CMake Error at CMakeLists.txt:60 (add_subdirectory):
  The source directory

    /app/third_party/abseil-cpp

  does not contain a CMakeLists.txt file.
```

joka921 commented:
`git submodule init`
`git submodule update`
should Download abseil.
Did the merge Work without conflicts?

Hannah Bast <notifications@github.com> schrieb am Sa., 11. Jan. 2020, 09:12:

> I just tried to build a version of the current master (includes the
> unicode upgrade) with this PR merged locally and encountered this error
> while building the docker container:
>
> CMake Error at CMakeLists.txt:60 (add_subdirectory):
>   The source directory
>
>     /app/third_party/abseil-cpp
>
>   does not contain a CMakeLists.txt file.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ad-freiburg/QLever/pull/302?email_source=notifications&email_token=ADS6SASFZKCKQ663O6WA5I3Q5F5QDA5CNFSM4KEKWXE2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIV4M2Q#issuecomment-573294186>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADS6SATV2RHIZOSOTKB57W3Q5F5QDANCNFSM4KEKWXEQ>
> .
>

hannahbast commented:
@joka921 In fact, it didn't, sorry for the confusion. Below you the see the list of the conflicting files. So I will wait until you have modified this pull request to be mergeable with the current master.

```
bast@galera:QLever$ git merge master
Auto-merging test/VocabularyGeneratorTest.cpp
CONFLICT (content): Merge conflict in test/VocabularyGeneratorTest.cpp
Auto-merging test/CMakeLists.txt
CONFLICT (content): Merge conflict in test/CMakeLists.txt
Auto-merging src/index/VocabularyGeneratorImpl.h
CONFLICT (content): Merge conflict in src/index/VocabularyGeneratorImpl.h
Auto-merging src/index/VocabularyGenerator.h
CONFLICT (content): Merge conflict in src/index/VocabularyGenerator.h
Auto-merging src/index/Index.h
CONFLICT (content): Merge conflict in src/index/Index.h
Auto-merging src/index/Index.cpp
CONFLICT (content): Merge conflict in src/index/Index.cpp
Auto-merging src/index/CMakeLists.txt
CONFLICT (content): Merge conflict in src/index/CMakeLists.txt
Auto-merging CMakeLists.txt
Automatic merge failed; fix conflicts and then commit the result.
```

hannahbast commented:
Awesome, I have started a build with this code at 9:39 h today, see my WA message. Fingers crossed

hannahbast commented:
@joka921 A minor detail: can you please call the intermediate files

```...partial-vocabulary-012``` instead of ```...partial-vocabulary12```
```...partial-ids-mmap-018``` instead of ```...partial-ids-mmap18```
```...compression-index...``` instead of ```...compression_index...```

Note that this entails **three** changes: (1) a dash before the final number, (2) fixed-width formatting of the number, so that the lexicographic order makes more sense when seeing them in a directory listing, (3) a dash instead of an underscore. If you feel that a fixed width of three is not enough, you can also make it four.

joka921 commented:
I will make this optionally. Without Level 5 there can be literals that
compare equal but are different. This leads to Index builds that behave
equally, but Producer different Indices in Case two of those equal literals
are sorted in a different Order in the vocabulary. I Wang to have
reproducible Index builds at least optionally as a Check dir Changes in the
Index builder. Thema I can simply Run cmp in the Index Files to See If I
have broken anything.

Hannah Bast <notifications@github.com> schrieb am So., 12. Jan. 2020, 10:02:

> *@hannahbast* commented on this pull request.
>
> Didn't we say earlier that using the QUATERNARY level is enough and that
> using the IDENTICAL level would affect the performance negatively, or am I
> confusing something here? Here is the corresponding quote from the ICU
> documentation, which Niklas included in his comments on this PR on
> 2020-01-02 13:57, highlights added by me:
>
> "Identical Level: When all other levels are equal, the identical level is
> used as a tiebreaker. The Unicode code point values of the NFD form of each
> string are compared at this level, just in case there is no difference at
> levels 1-4 . For example, Hebrew cantillation marks are only distinguished
> at this level. *This level should be used sparingly*, as only code point
> values differences between two strings is an extremely rare occurrence. *Using
> this level substantially decreases the performance* for both incremental
> comparison and sort key generation (as well as increasing the sort key
> length). It is also known as level 5 strength."
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ad-freiburg/QLever/pull/302?email_source=notifications&email_token=ADS6SAV7WWX4M767IKFJMATQ5LMDJA5CNFSM4KEKWXE2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRN5FSY#pullrequestreview-341562059>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADS6SAQT3BQEIPAP7LUTHBLQ5LMDJANCNFSM4KEKWXEQ>
> .
>

hannahbast commented:
@joka921 Travis complains that

```
Error: The source file ./src/util/BatchedPipeline.h does not match the code style
Use clang-format with the .clang-format provided in the QLever
repository's root to ensure all code files are formatted properly. We currently use the clang-format 8.0.0-3
```

niklas88 commented on src/index/ConstantsIndexCreation.h:
```diff
@@ -53,3 +53,7 @@ static const std::string PARTIAL_MMAP_IDS = ".partial-ids-mmap";
 static const std::string TMP_BASENAME_COMPRESSION = ".tmp.compression_index";
+
+// _________________________________________________________________
+// TODO: Comment
```
yeah I guess you're right

niklas88 commented on src/index/VocabularyGeneratorImpl.h:
```diff
@@ -215,37 +215,55 @@ void VocabularyMerger::doActualWrite(
-    } else {
-      std::sort(begin(els), end(els), pred);
+// ____________________________________________________________________________________________________________
+absl::flat_hash_map<Id, Id> createInternalMapping(std::vector<std::pair<string, Id>>* elsPtr) {
```
I'd like to stick to just one hash map implementation in normal QLever code (i.e. in libraries it's ok). So I think we should make `util::HashMap` use `absl::flat_hash_map` I think I once refactored it so that this shouldn't be too hard. What do you think?

joka921 replied:
For this PR I removed `absl` again, since the parallelism takes care of my current performance issues and we can do the change in a clean separate PR.

niklas88 commented on test/BatchedPipelineTest.cpp:
```diff
@@ -0,0 +1,186 @@
+          }
+          return std::nullopt;
+        },
+        [a = int(0)](const auto& x) mutable {
```
The use of `int(0)` is inconsistent to the above use of `[i = 0]`

joka921 commented on src/index/Index.cpp:
```diff
@@ -235,8 +256,11 @@ VocabularyData Index::passFileForVocabulary(const string& filename,
     VocabularyMerger v;
-    mergeRes =
-        v.mergeVocabulary(_onDiskBase, numFiles, _vocab.getCaseComparator());
+    auto identicalPred = [c = _vocab.getCaseComparator()](const auto& a,
```
I changed it, but can you point out any source where making lambdas `const auto` helps performance. Even the `all things constexpr` guys seem to always use plain `auto` for lambdas.
The only thing that you can prevent that way, is moving the lambda out, but typically compilers see through the lambda very well and the call operator of lambdas is always `const` by default, so I am not convinced that this substantially helps the code.

niklas88 replied:
Hmm, good point I guess you're right

joka921 commented on src/util/BatchedPipeline.h:
```diff
@@ -0,0 +1,636 @@
+
+/**
+ * @brief setup a pipeline that efficiently creates and transforms values. The
+ * Concurrency is used between the different levels
```
I chose stages.

joka921 commented on src/util/BatchedPipeline.h:
```diff
@@ -0,0 +1,636 @@
+}
+
+/**
+ * @brief setup a pipeline that efficiently creates and transforms values.
```
This comment is substantially different and only describes the differences.

joka921 commented on src/util/BatchedPipeline.h:
```diff
@@ -0,0 +1,636 @@
+namespace detail {
+
+/* This is used as a return value for the produceBatch calls of our pipeline
+ * elements If the first element (the bool) is false this means that our
```
The name is wrong. it should be `isExhausted` nice catch. This was a classical refactoring error.

niklas88 replied:
I agree, I'm not a fan of the underscores for private members either

joka921 commented on src/util/BatchedPipeline.h:
```diff
@@ -0,0 +1,636 @@
+ * pipeline is exhausted and there is no point in asking it for further batches.
+ */
+template <class T>
+struct VecT {
```
fixed.

joka921 commented on src/util/BatchedPipeline.h:
```diff
@@ -0,0 +1,636 @@
+
+/*
+ * This class represents an intermediate step of the Pipeline.
+ * It gets batches from the Producer and Applies the Transformer(s) (see below)
```
The Producer is more neutrally the `PreviousStage` of the Pipeline, would you like that better?

niklas88 replied:
I like Producer, I just think it can be a bit confusing that earlier comments mention a Creator which to me sounds like the same thing

joka921 commented on src/util/BatchedPipeline.h:
```diff
@@ -0,0 +1,636 @@
+  // _____________________________________________________________________
+  VecT<ResT> produceBatch() {
+    try {
+      auto res = _fut.get();
```
The problem here is that this function should be called `pickupBatch`.
It means "The next stage of the pipeline wants to have another batch. If the creation of the current batch is not yet finished, wait for it. When your current batch is picked up, you can start creating a new one using `orderNextBatch`.

Does this make more sense to you?

niklas88 replied:
Yes, thanks for the clarification

joka921 commented on src/util/BatchedPipeline.h:
```diff
@@ -0,0 +1,636 @@
+  return setupParallelPipelineRecursive<Parallelisms...>(
+      makeBatchedPipeline<NextParallelism>(std::forward<SoFar>(sofar),
+                                           std::forward<NextTransformer>(next)),
+      std::forward<MoreTransformers>(transformers)...);
```
The formatting is absolutely fine. the call the `makeBatchedPipeline` first takes the `sofar` and the `next` and appends one more step of our pipeline. Then we recurse on the extended pipeline and the remaining transformers.

joka921 commented on src/util/TupleHelpers.h:
```diff
@@ -0,0 +1,111 @@
+namespace detail {
+/* Implementation of setupTupleFromCallable (see below)
+ * Needed because we need the Pattern matching on the index_sequence
+ * TODO<joka921> In C++ 20 this could be done in place with templated
```
I highly doubt it since the standard is feature freezed but not yet published.

joka921 commented on src/index/StringSortComparator.h:
```diff
@@ -205,6 +207,23 @@ class LocaleManager {
+   * @param input The String to be normalized. Must be UTF-8 encoded
+   * @return The NFC canonical form of NFC in UTF-8 encoding.
+   */
+  std::string normalizeUtf8(std::string_view input) const {
```
I opened Issue #313 for this so I don't forget it. I do not want to do this while I still have PRs open that modify it to not get into rebasing hell. Otherwise I agree.

joka921 commented:
@hannahbast Sorry for the internal and meaningsless commit messages, I used this branch to transport content from my local machine to Galera.

@hannahbast @niklas88 
I have integrated Niklas' suggestions yesterday. Namely I have added comments and refactored the large lambdas out.

I have decided to integrate two more changes into this PR, they are separate commits and thus they should be easy to review separately.

1. We create SortKeys for all Vocabulary Elements on first sight, so that the actual sorting becomes cheaper. Otherwise the Sorting using ICU seems to be the bottleneck during IndexBuilding.

2. I replaced the `ad_utility::HashMap` Implementation with `absl::flat_hash_map` (previously `google::sparsehash`.) This is much faster and is necessary to get an actual speedup using this PR.

All in all, maybe @hannahbast should try building an index using this PR, and @niklas88 could have a look at the two additional changes.

niklas88 commented:
@joka921 the code looks good but the commit for replacing `google:sparsehash` the `QueryPlannerTest` faileds in Travis. I suspect this is the same hash order dependency problem that broke that test on s390x (issue #294). I don't see the fix to the test but the last commit is green so I'm worrying that this is only by luck.

niklas88 commented:
@joka921 I actually started on fixing that damn `QueryPlannerTest` but is was just too frustrating for the free time coding work I'm doing on QLever at the moment. I'm sorry this now falls on your feet. Honestly I'd be fine with just scrapping the half of that test that just compare to expected trees which is super flaky to begin with.

joka921 commented:
@niklas88 @hannahbast The issue of the QueryPlannerTest seemed fixable to me (Whenever the QueryPlanner can choose between equivalent Trees, we break the tie according to the CacheKey.

However implementing this I stumbled upon a probably serious issue: #314 
Which I will deal with after at least a night's sleep.

joka921 commented:
@niklas88 I applied a fix for the queryPlannerTest business. Let me know what you think of it (last commit) and verify if this indeed fixes the problem (I don't have that many machines to my disposal with different architectures).

niklas88 commented:
@joka921 the build currently fails both on Travis and locally with GCC 9.2.x

niklas88 commented:
`util/HashSet.h` uses Google dense_hashset, do you know if there is also an absl equivalent then we could remove the sparsehash dependency

niklas88 commented:
On the other hand absl uses `x86_64` specific options when the compiler is detected as GNU so I'll definitely have to report a bug there

joka921 commented:
@niklas88 there is `absl::flat_hash_set` which I am happy to integrate.
What do you want to file a bug for? I have not found any strange behavior, does something not work on your special Mainframe ISAs?

niklas88 commented:
Ok so I'm still getting a test failure in the `QueryPlannerTest` that looks like some order dependence (log attached). I guess this could be triggered by something other than hash map order, maybe endianess as s390x is basically the last pure Big Endian arch?

The abseil CMake magic used `-maes -msse4.2` whenever it detected a GNU compiler with the current absl master branch this is fixed. Our special Mainframe ISA does indeed also have in CPU crypto but I'd already be happy with not trying to use SSE :dagger: 
[failing_test.log](https://github.com/ad-freiburg/QLever/files/4124707/failing_test.log)

joka921 commented:
I just spent some time running a debugger on the first of your failing unit tests. 
At first glance this is NOT and ordering problem, since the Execution Tree chosen by your machine is more expensive than the expected one. So I assume that some of the `CostEstimates` are platform dependent. (Hopefully only the dummy ones used during unit tests when there is no `QueryExecutionContext`). I'll sent you a verbosity patch that outputs all the sizeEstimates tomorrow, then we can compare, where the difference might come from.

joka921 commented:
@niklas88 
Ok so I just pushed a commit that verbosely outputs all the cost estimates of all subtrees.
Could you please run the QueryPlannerTest and send me the output so I can compare, where the differences lie in the estimates;

niklas88 commented:
[test_fail.txt](https://github.com/ad-freiburg/QLever/files/4129516/test_fail.txt)
Do you have any objections to setting the abseil subrepository to the current HEAD so we get that build fix?

joka921 commented:
@niklas88 
Ok, so now this should work. If not, please send me the same file as yesterday (this time you have to build with ` -DLOGLEVEL=TRACE` to obtain the verbose output.
In case it does work, have a glance at the code of the last two commits and feel free to finally merge this

niklas88 commented:
Great work! The last commit also now fixes #294 and make all tests pass on s390x including the E2E Tests
